### ISSUE TITLE
“Replace recovery codes” button does not work on iOS

### ISSUE THREAD

**Issue by vasilicatamas** [mozilla/fxa#5629]
**Build**
Prod [Train 1.170.0]

**Affected platforms**
iOS 13.4.1 

**Steps to reproduce**
1. Login using a Firefox account that has 2FA enabled.
2. Go to https://accounts.firefox.com/settings
3. Tap on "Change" button from 2FA section.
4. Tap on “Replace recovery codes” button.

**Expected results**
“Save recovery codes” modal is displayed.

**Actual Results**
Nothing happens when tapping on “Replace recovery codes” button.

![issue8](https://user-images.githubusercontent.com/18114210/84291666-1832c280-ab4e-11ea-9ccd-6e8379c6b929.gif)

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/FXA-2088)
┆Issue Number: FXA-2088

**Comment by vcarciu** [comment]
I tried to reproduce the issue on iOS 13.5.1, with FF for iOS version 27.0(18428) and I was able to constantly reproduce the problem as described by the reporter.

**Comment by vcarciu** [comment]
After talking to Ica, I tried to reproduce the issue on Safari and Chrome, using an iPhone XS device, with iOS 13.5.1 . On both Safari and Chrome, the “Save recovery codes” modal is displayed without issues. 

I also tried again on FF for iOS and I am constantly reproducing the issue. 

For any other info, please leave a comment here and I will try to help if possible.

Thank you.

**Comment by LZoog** [comment]
@vcarciu - thanks for looking into that. @jodyheavener and I discovered something even stranger this morning. He could reproduce the issue, I could not, so we were going through some various scenarios on his physical device to try to isolate the bug.

He clicked "disable" to remove two step auth, then went through the flow from the start. After resetting the flow like this, he could not reproduce any longer. If you "disable" two step auth, re-enable, _then_ click "replace recovery codes," does it work?

If it does, this _might_ suggest that this is not actually iOS specific but that this just happened to occur on two iOS devices. I wonder if it's possible that in some rare edge cases, the current codes somehow become "stuck" and our `replaceRecoveryCodes` method fails for some reason. The fact that it doesn't occur reliably in Safari, FF for Android, XCode, or across all iOS devices (I cannot reproduce on my iPhone X, Wil cannot reproduce on his iPhone SE) makes this very difficult to debug because we can't check the console or network tabs. If it reliably works after disabling then re-enabling the codes, maybe we can figure out how to reproduce it on other devices.

**Comment by vcarciu** [comment]
@LZoog : Nice finding ! Indeed, after disable\enable , the “Save recovery codes” modal is displayed.

**Comment by LZoog** [comment]
Here’s the latest status on this issue:

TLDR; the steps to reproduce this issue are unknown and I cannot reproduce in various devices or in XCode, and I can't find a logged error for it. @vcarciu I think the best plan of action here would be to ask you to try to find reliable steps to reproduce this - if we can do that, we can better work towards solving it. Even if we don't have the steps down, having a device with the issue could be helpful to work with you to ask you to trigger the button and immediately check Sentry's latest logs see if there is an error buried in Sentry somewhere. I suspect it's not an issue only on iOS, but that could be incorrect. It's not an issue with a user's account since it's only an issue on FF and not Safari (which is weird) or Chrome.

Of the engineers that have tried to reproduce this issue, none have been successful except for Jody on a physical iPhone. Since disabling 2FA and re-enabling it however, he can no longer reproduce it.

I’ve searched through the Sentry logs and do see a very small number of events tied to `two_step_authentication`, which appears to be the start of the flow that this issue would occur in (`_replaceRecoveryCodes` in particular), but I don’t see any errors from an iOS device triggered when QA verified this was reproducible. It’s possible someone with better Sentry grepping knowledge could find something I didn’t, but I also searched for errors reported at the `/settings/two_step_authentication` URL with `os:"iOS 13.5.1"` specified, with events happening on Tuesday the 30th (I eyeballed this, I only see options for last event/first event) on train `1.177.0`/`1.177.1` (a point release was done this day). There should be multiple events for this error if it is indeed hidden in Sentry somewhere because Jody clicked on the button in question multiple times to reproduced the issue and QA confirmed again on the 30th that this was an issue. In hindsight, I wish I'd worked with Jody to trigger the problem and immediately view latest Sentry logs to try to find a connection just in case there is something buried in there. If we have a device where this is reproducible perhaps we can still do this.

Thankfully this problem appears uncommon, but I can't claim it's an isolated edgecase that we can ignore as QA was able to reproduce this on two devices and Jody on one, but I’ve kind of hit a dead end here since I can't reproduce or find a logged error.

**Comment by vasilicatamas** [comment]
I just noticed that the only scenario when the modal is displayed is when the “Replace recovery codes” link is clicked just after enabling the 2FA using the iOS device. But the issue can be reproduced after logout and login. 

@LZoog Hope that this info will help you to reproduce this bug.

**Comment by data-sync-user** [comment]
➤ Vasilica Tamas commented:

This issue can be closed because it is no longer applicable, considering the New Settings page replacement.
